common/src/auth/SharedAccessSignature.cs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Text;
+using System.Net;
+
+namespace Microsoft.Azure.Devices.Common.Authorization
+{
+    public class SharedAccessSignature
+    {
+        public SharedAccessSignature(string sharedAccessSignatureName, DateTime expiresOn, string expiry, string keyName, string signature, string encodedAudience)
+        {
+            if (string.IsNullOrWhiteSpace(sharedAccessSignatureName))
+            {
+                throw new ArgumentNullException(nameof(sharedAccessSignatureName));
+            }
+
+            ExpiresOnUtc = expiresOn;
+
+            if (IsExpired())
+            {
+                throw new UnauthorizedAccessException("The specified SAS token is expired");
+            }
+        }
+
+        public DateTime ExpiresOnUtc
+        {
+            get;
+            private set;
+        }
+
+        public static SharedAccessSignature Parse(string shareAccessSignatureName, string rawToken)
+        {
+            if (string.IsNullOrWhiteSpace(shareAccessSignatureName))
+            {
+                throw new ArgumentNullException(nameof(shareAccessSignatureName));
+            }
+
+            if (string.IsNullOrWhiteSpace(rawToken))
+            {
+                throw new ArgumentNullException(nameof(rawToken));
+            }
+
+            IDictionary<string, string> parsedFields = ExtractFieldValues(rawToken);
+
+            string signature;
+            if (!parsedFields.TryGetValue(SharedAccessSignatureConstants.SignatureFieldName, out signature))
```

Comment:
all these can be inline type declarations, e.g. `parsedFields.TryGetValue(..., out string signature);`